File: faos_core/connectors/webhooks.ts
/**
 * FAOS v6.0 — External webhook / delivery connector
 */

export type WebhookPayload = {
  event: string;
  pipeline_id: string;
  brand?: string;
  body: Record<string, unknown>;
};

export type WebhookDeliveryResult = {
  ok: boolean;
  event: string;
  pipeline_id: string;
  delivered_at: string;
  mode: "dry_run" | "http";
  status_code?: number;
  error?: string;
};

/**
 * Deliver approval-gate output. Without WEBHOOK_URL, dry-runs safely.
 */
export async function deliverWebhook(
  payload: WebhookPayload,
  options?: { url?: string; dry_run?: boolean }
): Promise<WebhookDeliveryResult> {
  const url = options?.url || process.env.FAOS_WEBHOOK_URL?.trim();
  const dry = options?.dry_run ?? !url;

  if (dry || !url) {
    return {
      ok: true,
      event: payload.event,
      pipeline_id: payload.pipeline_id,
      delivered_at: new Date().toISOString(),
      mode: "dry_run",
    };
  }

  try {
    const res = await fetch(url, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({
        ...payload,
        source: "faos_v6",
        timestamp: new Date().toISOString(),
      }),
    });
    return {
      ok: res.ok,
      event: payload.event,
      pipeline_id: payload.pipeline_id,
      delivered_at: new Date().toISOString(),
      mode: "http",
      status_code: res.status,
      error: res.ok ? undefined : `http_${res.status}`,
    };
  } catch (err) {
    return {
      ok: false,
      event: payload.event,
      pipeline_id: payload.pipeline_id,
      delivered_at: new Date().toISOString(),
      mode: "http",
      error: err instanceof Error ? err.message : "webhook_failed",
    };
  }
}

File: backend/faos_core/connectors/webhooks.py
"""FAOS v6.0 — External webhook / delivery connector (Python)."""

from __future__ import annotations

import asyncio
import json
import os
import urllib.error
import urllib.request
from datetime import datetime, timezone
from typing import Any, Dict, Optional


def _post_sync(url: str, payload: Dict[str, Any]) -> int:
    data = json.dumps(payload).encode("utf-8")
    req = urllib.request.Request(
        url,
        data=data,
        headers={"Content-Type": "application/json"},
        method="POST",
    )
    with urllib.request.urlopen(req, timeout=15) as res:  # noqa: S310
        return int(res.status)


async def deliver_webhook(
    payload: Dict[str, Any],
    *,
    url: Optional[str] = None,
    dry_run: Optional[bool] = None,
) -> Dict[str, Any]:
    target = (url or os.getenv("FAOS_WEBHOOK_URL") or "").strip()
    dry = dry_run if dry_run is not None else not bool(target)

    if dry or not target:
        return {
            "ok": True,
            "event": payload.get("event"),
            "pipeline_id": payload.get("pipeline_id"),
            "delivered_at": datetime.now(timezone.utc).isoformat(),
            "mode": "dry_run",
        }

    body = {
        **payload,
        "source": "faos_v6",
        "timestamp": datetime.now(timezone.utc).isoformat(),
    }

    try:
        status = await asyncio.to_thread(_post_sync, target, body)
        ok = 200 <= status < 300
        return {
            "ok": ok,
            "event": payload.get("event"),
            "pipeline_id": payload.get("pipeline_id"),
            "delivered_at": datetime.now(timezone.utc).isoformat(),
            "mode": "http",
            "status_code": status,
            "error": None if ok else f"http_{status}",
        }
    except Exception as exc:  # noqa: BLE001
        return {
            "ok": False,
            "event": payload.get("event"),
            "pipeline_id": payload.get("pipeline_id"),
            "delivered_at": datetime.now(timezone.utc).isoformat(),
            "mode": "http",
            "error": str(exc),
        }
